build fails randomly

**Reproduce**

- `rm -rf /var/lib/docker`
- compile and install latest **master** docker (I'm on latest fedora 23)
- from inside the repo: `docker build -t dry-run-test .`
- watch it fails at lvm2 compile step or also at update (it's random...)

When docker build with docker 1.10.1 everything runs fine. I narrowed down the issue, it seems to fail trying to compilie lvm2 (there's also a data race in daemon logs).

**Info**
```

[...truncated build output...]

checking whether to enable readline... maybe
checking whether to enable realtime support... 
checking whether to enable OCF resource agents... no
checking whether to build cluster LVM daemon... none
checking whether to build cluster mirror log daemon... no
checking whether to enable debugging... no
checking for C optimisation flag... -O2
checking whether to gather gcov profiling data... no
checking whether to enable unit testing... no
checking whether to enable valgrind awareness of pools... no
checking whether to use device-mapper... yes
checking whether to build LVMetaD... no
checking whether to enable synchronisation with udev processing... no
checking whether to enable installation of udev rules required for synchronisation... no
checking whether to enable executable path detection in udev rules... no
checking whether udev supports built-in blkid... checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for PKGCONFIGINIT... no
pkg-config initialized
no
checking whether to enable O_DIRECT... yes
checking whether to build liblvm2app.so application library... no
checking whether to compile liblvm2cmd.so... no
checking whether to build Python wrapper for liblvm2app.so... no
checking whether to install fsadm... yes
checking whether to install blkdeactivate... yes
checking whether to use dmeventd... 
checking for getline in -lc... yes
checking for canonicalize_file_name in -lc... yes
checking for dlopen in -ldl... yes
checking whether to enable selinux support... yes
checking for sepol_check_context in -lsepol... no
checking for is_selinux_enabled in -lselinux... no
checking for getopt.h... (cached) yes
configure: WARNING: Disabling selinux
docker run --rm -i --privileged -e BUILDFLAGS -e KEEPBUNDLE -e DOCKER_BUILD_GOGC -e DOCKER_BUILD_PKGS -e DOCKER_CLIENTONLY -e DOCKER_DEBUG -e DOCKER_EXPERIMENTAL -e DOCKERFILE -e DOCKER_GRAPHDRIVER -e DOCKER_INCREMENTAL_BINARY -e DOCKER_REMAP_ROOT -e DOCKER_STORAGE_OPTS -e DOCKER_USERLANDPROXY -e TESTDIRS -e TESTFLAGS -e TIMEOUT -v "/home/runcom/koding/docker/bundles:/go/src/github.com/docker/docker/bundles" -t "docker-dev:master" hack/make.sh dynbinary test-integration-cli
Unable to find image 'docker-dev:master' locally
Pulling repository docker.io/library/docker-dev
docker: Tag master not found in repository docker.io/library/docker-dev.
See 'docker run --help'.
Makefile:123: recipe for target 'test-integration-cli' failed
make: *** [test-integration-cli] Error 125
```
*daemon logs*
```
Feb 19 11:35:11 fedora docker[14686]: checking for getline in -lc... yes
Feb 19 11:35:11 fedora docker[14686]: checking for canonicalize_file_name in -lc... yes
Feb 19 11:35:11 fedora docker[14686]: checking for dlopen in -ldl... yes
Feb 19 11:35:11 fedora docker[14686]: checking whether to enable selinux support... yes
Feb 19 11:35:11 fedora docker[14686]: checking for sepol_check_context in -lsepol... no
Feb 19 11:35:11 fedora docker[14686]: configure: WARNING: Disabling selinux
Feb 19 11:35:11 fedora docker[14686]: checking for is_selinux_enabled in -lselinux... no
Feb 19 11:35:11 fedora docker[14686]: checking for getopt.h... (cached) yes
Feb 19 11:35:11 fedora docker[14686]: ==================
Feb 19 11:35:11 fedora docker[14686]: WARNING: DATA RACE
Feb 19 11:35:11 fedora docker[14686]: Read by goroutine 84:
Feb 19 11:35:11 fedora docker[14686]: bufio.(*Writer).Write()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/bufio/bufio.go:589 +0x9b
Feb 19 11:35:11 fedora docker[14686]: bufio.(*ReadWriter).Write()
Feb 19 11:35:11 fedora docker[14686]: <autogenerated>:20 +0x8e
Feb 19 11:35:11 fedora docker[14686]: fmt.Fprintf()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/fmt/print.go:189 +0xd2
Feb 19 11:35:11 fedora docker[14686]: net/http.(*chunkWriter).Write()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/net/http/server.go:268 +0x345
Feb 19 11:35:11 fedora docker[14686]: bufio.(*Writer).flush()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/bufio/bufio.go:562 +0x1a7
Feb 19 11:35:11 fedora docker[14686]: bufio.(*Writer).Flush()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/bufio/bufio.go:551 +0x3a
Feb 19 11:35:11 fedora docker[14686]: net/http.(*response).Flush()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/net/http/server.go:1204 +0x7c
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/pkg/ioutils.(*WriteFlusher).Flush()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/pkg/ioutils/writeflusher.go:49 +0x104
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/pkg/ioutils.(*WriteFlusher).Write()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/pkg/ioutils/writeflusher.go:34 +0x151
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/pkg/streamformatter.(*StderrFormatter).Write()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/pkg/streamformatter/streamformatter.go:167 +0x17e
Feb 19 11:35:11 fedora docker[14686]: io.copyBuffer()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/io/io.go:383 +0x303
Feb 19 11:35:11 fedora docker[14686]: io.Copy()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/io/io.go:351 +0x78
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/container.AttachStreams.func2()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/container/container.go:444 +0x26d
Feb 19 11:35:11 fedora docker[14686]: Previous write by goroutine 65:
Feb 19 11:35:11 fedora docker[14686]: bufio.(*Writer).Write()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/bufio/bufio.go:607 +0x4ce
Feb 19 11:35:11 fedora docker[14686]: bufio.(*ReadWriter).Write()
Feb 19 11:35:11 fedora docker[14686]: <autogenerated>:20 +0x8e
Feb 19 11:35:11 fedora docker[14686]: fmt.Fprintf()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/fmt/print.go:189 +0xd2
Feb 19 11:35:11 fedora docker[14686]: net/http.(*chunkWriter).Write()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/net/http/server.go:268 +0x345
Feb 19 11:35:11 fedora docker[14686]: bufio.(*Writer).flush()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/bufio/bufio.go:562 +0x1a7
Feb 19 11:35:11 fedora docker[14686]: bufio.(*Writer).Flush()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/bufio/bufio.go:551 +0x3a
Feb 19 11:35:11 fedora docker[14686]: net/http.(*response).Flush()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/net/http/server.go:1204 +0x7c
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/pkg/ioutils.(*WriteFlusher).Flush()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/pkg/ioutils/writeflusher.go:49 +0x104
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/pkg/ioutils.(*WriteFlusher).Write()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/pkg/ioutils/writeflusher.go:34 +0x151
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/pkg/streamformatter.(*StdoutFormatter).Write()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/pkg/streamformatter/streamformatter.go:152 +0x127
Feb 19 11:35:11 fedora docker[14686]: io.copyBuffer()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/io/io.go:383 +0x303
Feb 19 11:35:11 fedora docker[14686]: io.Copy()
Feb 19 11:35:11 fedora docker[14686]: /usr/lib/golang/src/io/io.go:351 +0x78
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/container.AttachStreams.func2()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/container/container.go:444 +0x26d
Feb 19 11:35:11 fedora docker[14686]: Goroutine 84 (running) created at:
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/container.AttachStreams()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/container/container.go:455 +0x4aa
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/container.(*Container).Attach()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/container/container.go:362 +0x172
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/daemon.(*Daemon).containerAttach()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/daemon/attach.go:111 +0x5f8
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/daemon.(*Daemon).ContainerAttachRaw()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/daemon/attach.go:62 +0x12e
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/builder/dockerfile.(*Builder).run.func1()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/builder/dockerfile/internals.go:543 +0xf3
Feb 19 11:35:11 fedora docker[14686]: Goroutine 65 (running) created at:
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/container.AttachStreams()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/container/container.go:454 +0x448
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/container.(*Container).Attach()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/container/container.go:362 +0x172
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/daemon.(*Daemon).containerAttach()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/daemon/attach.go:111 +0x5f8
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/daemon.(*Daemon).ContainerAttachRaw()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/daemon/attach.go:62 +0x12e
Feb 19 11:35:11 fedora docker[14686]: github.com/docker/docker/builder/dockerfile.(*Builder).run.func1()
Feb 19 11:35:11 fedora docker[14686]: /home/runcom/koding/docker/.gopath/src/github.com/docker/docker/builder/dockerfile/internals.go:543 +0xf3
Feb 19 11:35:11 fedora docker[14686]: ==================
Feb 19 11:35:11 fedora docker[14686]: time="2016-02-19T11:35:11.615221595+01:00" level=debug msg="Build cancelled, killing and removing container: 9fa0323fa1c59d06afd7118b90e33e49f9a06e233772620333175fede3353969"
Feb 19 11:35:11 fedora docker[14686]: time="2016-02-19T11:35:11.615478096+01:00" level=debug msg="Sending 9 to 9fa0323fa1c59d06afd7118b90e33e49f9a06e233772620333175fede3353969"
Feb 19 11:35:11 fedora docker[14686]: checking for library containing tgetent... 
Feb 19 11:35:11 fedora docker[14686]: time="2016-02-19T11:35:11.648425438+01:00" level=debug msg="attach: stderr: end"
Feb 19 11:35:11 fedora docker[14686]: time="2016-02-19T11:35:11.648524391+01:00" level=debug msg="attach: stdout: end"
Feb 19 11:35:11 fedora docker[14686]: time="2016-02-19T11:35:11.717904775+01:00" level=debug msg="Releasing addresses for endpoint insane_davinci's interface on network bridge"
Feb 19 11:35:11 fedora docker[14686]: time="2016-02-19T11:35:11.718030376+01:00" level=debug msg="ReleaseAddress(LocalDefault/172.17.0.0/16, 172.17.0.2)"
Feb 19 11:35:11 fedora docker[14686]: time="2016-02-19T11:35:11.771204653+01:00" level=warning msg="could not write error response: write unix /var/run/docker.sock->@: write: broken pipe"
```

I got this also after/during the apt-get update command in the Dockerfile :confused: 
```
Get:197 http://httpredir.debian.org/debian/ jessie/main yum all 3.4.3-3 [620 kB]
Get:198 http://httpredir.debian.org/debian/ jessie/main createrepo all 0.10.3-1 [48.8 kB]
Get:199 http://httpredir.debian.org/debian/ jessie/main mingw-w64-common all 3.2.0-2 [3713 kB]
Get:200 http://security.debian.org/ jessie/updates/main openssl amd64 1.0.1k-3+deb8u4 [677 kB]
Get:201 http://httpredir.debian.org/debian/ jessie/main mingw-w64-i686-dev all 3.2.0-2 [1840 kB]
Get:202 http://httpredir.debian.org/debian/ jessie/main gcc-mingw-w64-i686 amd64 4.9.1-19+14.3 [23.5 MB]
Get:203 http://security.debian.org/ jessie/updates/main curl amd64 7.38.0-4+deb8u3 [200 kB]
Get:204 http://httpredir.debian.org/debian/ jessie/main mingw-w64-x86-64-dev all 3.2.0-2 [2883 kB]
docker run --rm -i --privileged -e BUILDFLAGS -e KEEPBUNDLE -e DOCKER_BUILD_GOGC -e DOCKER_BUILD_PKGS -e DOCKER_CLIENTONLY -e DOCKER_DEBUG -e DOCKER_EXPERIMENTAL -e DOCKERFILE -e DOCKER_GRAPHDRIVER -e DOCKER_INCREMENTAL_BINARY -e DOCKER_REMAP_ROOT -e DOCKER_STORAGE_OPTS -e DOCKER_USERLANDPROXY -e TESTDIRS -e TESTFLAGS -e TIMEOUT -v "/home/runcom/koding/docker/bundles:/go/src/github.com/docker/docker/bundles" -t "docker-dev:fix-creds-store" hack/make.sh test-unit
Unable to find image 'docker-dev:fix-creds-store' locally
Pulling repository docker.io/library/docker-dev
docker: Tag fix-creds-store not found in repository docker.io/library/docker-dev.
See 'docker run --help'.
Makefile:126: recipe for target 'test-unit' failed
make: *** [test-unit] Error 125
```

I'm hitting this bug with 1.11 always now - can anyone test this? I get this kind of error even with other projects:
```
  Downloading six-1.10.0-py2.py3-none-any.whl
Collecting MarkupSafe (from Jinja2>=2.4->Flask==0.10.1->docker-registry[bugsnag,newrelic,cors])
  Downloading MarkupSafe-0.23.tar.gz
Installing collected packages: backports.lzma, blinker, Werkzeug, MarkupSafe, Jinja2, itsdangerous, Flask, greenlet, gevent, gunicorn, PyYAML, requests, M2Crypto, sqlalchemy, webob, bugsnag, Six, Flask-cors, newrelic, docker-registry
  Running setup.py install for backports.lzma
  Running setup.py install for blinker
  Running setup.py install for MarkupSafe
  Running setup.py install for itsdangerous
  Running setup.py install for Flask
  Running setup.py install for greenlet
  Running setup.py install for gevent
  Running setup.py install for PyYAML
  Running setup.py install for M2Crypto
  Running setup.py install for sqlalchemy
  Running setup.py install for bugsnag
  Running setup.py install for Flask-cors
  Running setup.py install for newrelic
  Running setup.py install for docker-registry
Successfully installed Flask-0.10.1 Flask-cors-1.10.3 Jinja2-2.8 M2Crypto-0.22.3 MarkupSafe-0.23 PyYAML-3.11 Six-1.10.0 Werkzeug-0.11.4 backports.lzma-0.0.3 blinker-1.3 bugsnag-2.0.2 docker-registry-1.0.0.dev0 gevent-1.0.1 greenlet-0.4.9 gunicorn-19.1.1 itsdangerous-0.24 newrelic requests-2.3.0 sqlalchemy-0.9.4 webob-1.6.0
++ python -c 'import boto; import os; print os.path.dirname(boto.__file__)'
+ patch /usr/lib/python2.7/site-packages/boto/connection.py
docker run --rm -i  -t "skopeo-dev:enhanc" hack/make.sh test-integration
Unable to find image 'skopeo-dev:enhanc' locally
Pulling repository docker.io/library/skopeo-dev
docker: Error: image library/skopeo-dev not found.
See 'docker run --help'.
Makefile:55: recipe for target 'test-integration' failed
make: *** [test-integration] Error 125
```

--- a/api/server/router/build/build_routes.go
+++ b/api/server/router/build/build_routes.go
@@ -9,6 +9,7 @@ import (
 	"net/http"
 	"strconv"
 	"strings"
+	"sync"
 
 	"github.com/Sirupsen/logrus"
 	"github.com/docker/docker/api/server/httputils"
@@ -94,6 +95,18 @@ func newImageBuildOptions(ctx context.Context, r *http.Request) (*types.ImageBui
 	return options, nil
 }
 
+type syncWriter struct {
+	w  io.Writer
+	mu sync.Mutex
+}
+
+func (s *syncWriter) Write(b []byte) (count int, err error) {
+	s.mu.Lock()
+	count, err = s.w.Write(b)
+	s.mu.Unlock()
+	return
+}
+
 func (br *buildRouter) postBuild(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
 	var (
 		authConfigs        = map[string]types.AuthConfig{}
@@ -172,6 +185,7 @@ func (br *buildRouter) postBuild(ctx context.Context, w http.ResponseWriter, r *
 	if buildOptions.SuppressOutput {
 		out = notVerboseBuffer
 	}
+	out = &syncWriter{w: out}
 	stdout := &streamformatter.StdoutFormatter{Writer: out, StreamFormatter: sf}
 	stderr := &streamformatter.StderrFormatter{Writer: out, StreamFormatter: sf}
 

